gpio: stm32: initialise according to `zephyr,pm-device-runtime-auto`

JordanYates opened the pull request:
Don't automatically disable all GPIO ports just because `PM_DEVICE_RUNTIME` is enabled. Require the user to explicitly call `pm_device_runtime_enable` on the port, or add `zephyr,pm-device-runtime-auto` to the devicetree node.

JordanYates committed:
gpio: stm32: initialise according to `zephyr,pm-device-runtime-auto`

Don't automatically disable all GPIO ports just because
`PM_DEVICE_RUNTIME` is enabled. Require the user to explicitly call
`pm_device_runtime_enable` on the port, or add
`zephyr,pm-device-runtime-auto` to the devicetree node.

Signed-off-by: Jordan Yates <jordan@embeint.com>

mathieuchopstm commented on drivers/gpio/gpio_stm32.c:
```diff
@@ -688,13 +687,15 @@ static int gpio_stm32_pm_action(const struct device *dev,
 		return gpio_stm32_clock_request(dev, true);
 	case PM_DEVICE_ACTION_SUSPEND:
 		return gpio_stm32_clock_request(dev, false);
+	case PM_DEVICE_ACTION_TURN_OFF:
+	case PM_DEVICE_ACTION_TURN_ON:
+		break;
```
Doesn't look required for `pm_device_driver_init` to work, `-ENOTSUP` seems allowed:
https://github.com/zephyrproject-rtos/zephyr/blob/3d2059f9b66faf2b6cf4be15b2f4adc2f092f42b/subsys/pm/device.c#L371-L375

mathieuchopstm replied:
Nevermind. Looking through the framework, returning `-ENOTSUP` will break `pm_device_action_run`.

erwango reviewed (changes requested):
While I'm open on this change:
- this will break `samples/boards/st/power_mgmt/blinky/` which should be updated accordingly
- If there is a change of behavior expected from the application, this should be documented in migration guide.
- What are the impacts on devices making use of pinctrl which is based on gpio ? Today GPIO ports of pins are clocked off after transition. This won't be the case anymore w/o further changes in pinctrl driver (or pinctrl driver clients). Or I'm being optimistic and this is not working that way today ?

JordanYates commented:
So the only thing that PM controls is enabling or disabling the peripheral clock. As far as I can tell during my testing (on a STM32L451), the only observable consequence on disabling the clock is that all the register state is lost. There isn't any increase in power consumption in sleep modes (since the internal clocks are disabled anywhere), and presumably any additional power consumption in active modes is negligible compared to the mA of the CPU running.

Enabling PM on the port appears to me to be all downside, no upside. Without any reason to use PM on the port, disabling it by default should not break anything or require migration notes.

Obviously you have much deeper knowledge of STM parts than I do, have I missed some reason as to why PM on the GPIO port is advantageous?

erwango commented:
> Obviously you have much deeper knowledge of STM parts than I do, have I missed some reason as to why PM on the GPIO port is advantageous?

This is a fair question, so I took the time to get some feedback with our FAEs around this point. Disabling clocks is part of the general recommendation to achieve lowest possible figures in low power scenarios, even on recent series, even on field use cases (so not only a theoretical recommendation). Enabling clocks still drain current when enabled even if hardly measurable. It is also mentioned in [power management application note](https://www.st.com/resource/en/application_note/an5930-guidelines-for-power-management-on-stm32h5-mcus-stmicroelectronics.pdf).

JordanYates commented:
> This is a fair question, so I took the time to get some feedback with our FAEs around this point. Disabling clocks is part of the general recommendation to achieve lowest possible figures in low power scenarios, even on recent series, even on field use cases (so not only a theoretical recommendation). Enabling clocks still drain current when enabled even if hardly measurable. It is also mentioned in [power management application note](https://www.st.com/resource/en/application_note/an5930-guidelines-for-power-management-on-stm32h5-mcus-stmicroelectronics.pdf).

If I had to choose between "GPIO register configuration is lost" and "Fractional microamp due to GPIO clock gating", I know which one I would bet on as blowing out the power budget :). Since you agree that the drain is hardly measurable, I can't imagine that this change will be noticeable by users (except for the fact their GPIOs now keep their configured state) .

erwango commented:
> If I had to choose between "GPIO register configuration is lost" and "Fractional microamp due to GPIO clock gating", I know which one I would bet on as blowing out the power budget :). Since you agree that the drain is hardly measurable, I can't imagine that this change will be noticeable by users (except for the fact their GPIOs now keep their configured state) .

I fully understand this point of view. But note that the result of the observations you've done may vary across the whole STM32 family this driver is addressing.

JordanYates commented:
> I fully understand this point of view. But note that the result of the observations you've done may vary across the whole STM32 family this driver is addressing.

That's why PR's are assigned to STM experts who can identify any potential variations :).
Regardless, automatically enabling runtime PM on an instance is incorrect behavior from a driver.
From a conceptual point of view, the current setup does not provide a way to NOT enable runtime PM, while this PR allows it to be enabled or disabled as desired (through devicetree).

erwango commented:
> From a conceptual point of view, the current setup does not provide a way to NOT enable runtime PM, while this PR allows it to be enabled or disabled as desired (through devicetree).

To be clear: I'm fine with the change.
I'm just raising the following three points in order to avoid gotchas for existing users:
- `samples/boards/st/power_mgmt/blinky/` should be updated accordingly
- If there is a change of behavior expected from the application, this should be documented in migration guide.
- Regarding the 3rd point, this was raised on the top of my mind and I think we're ok now on this one. Let's discard

JordanYates commented:
> `samples/boards/st/power_mgmt/blinky/` should be updated accordingly

What changes do you want there?  `zephyr,pm-device-runtime-auto` added to each GPIO port on the 4 explicitly supported boards?

erwango commented:
> What changes do you want there? zephyr,pm-device-runtime-auto added to each GPIO port on the 4 explicitly supported boards?

What about calling `pm_device_runtime_enable()`? (I guess it should be possible to use a macro that would call it on all enabled GPIO ports)

JordanYates committed:
samples: st: power_mgmt: blinky: enable PM on GPIOs

Enable runtime PM on each GPIO port to preserve the legacy PM GPIO
behaviour for this sample.

Signed-off-by: Jordan Yates <jordan@embeint.com>
